src/newrelic.ts

```diff
@@ -40,53 +65,361 @@ function addSegment(spec: Spec) {
-  const seneca = this
+                            let peract = (context.peract[mi] = (context.peract[mi] || {}))
 
-  seneca.order.inward.add((spec: Spec) => addSegment(spec))
-  seneca.order.outward.add((spec: Spec) => endSegment(spec))
+                            peract.pattern = pattern
+                            peract.endSegment = endSegment
+                        },
+                        function endSegment() {
+                            console.log('end segment ' + segment)
+                        }
+                    )
+                }
+            }
+        },
+        endSegments: () => {
+            const mi = spec.data.meta.mi
+
+            if(spec.ctx.seneca.context.peract) {
+                const peract = spec.ctx.seneca.context.peract[mi]
+                if(peract) {
+                    (peract.endSegment())
+                }
+            }
+        },
+        extractFromSpec: () => {
+
+        }
+    }
 }
 
+function preload(this: any, opts: any) {
+  const seneca = this;
+  const { options }: { options: NewRelicOptions } = opts;
+  const segmentIsEnabled = options && options.segment && options.segment.enabled;
+  const tracingIsEnabled = options && options.tracing && options.tracing.enabled;
+  const metricsIsEnabled = options && options.metrics && options.metrics.enabled;
+  
+  let telemetryCollector: Nullable<ReturnType<typeof TelemetryCollector>> = null;
+  if (tracingIsEnabled && options.tracing) {
+    const { accountApiKey, serviceName, host } = options.tracing;
+    telemetryCollector = TelemetryCollector(
+      accountApiKey, serviceName, host
+    );
+  }
+
+  seneca.order.inward.add((spec: Spec) => {
+      if (segmentIsEnabled) {
+          changeSpec(spec, 'inward').addSegments();
+      }
+      if (telemetryCollector) {
+          const specMetadata = telemetryCollector.extractFromSpec(spec, 'inward');
+          telemetryCollector.updateSpecList(specMetadata);
+      }
+  })
+
+  seneca.order.outward.add((spec: Spec) => {
+      if (segmentIsEnabled) {
+          changeSpec(spec, 'outward').endSegments()
+      }
+      if (telemetryCollector) {
+          const specMetadata = telemetryCollector.extractFromSpec(spec, 'outward');
+          telemetryCollector.updateSpecList(specMetadata);
+      }
+  })
+
+  if (metricsIsEnabled) {
+    if (!options.metrics?.accountApiKey) {
+      throw new Error("Please provide accountApiKey parameter to Metrics API");
+    }
+    const metricsClient = new MetricClient({
+      apiKey: options.metrics.accountApiKey,
+    })
 
-function newrelic(this: any, _options: any) {
-  // const seneca: any = this
+    this.metricsClient = metricsClient;
+  }
 }
 
+function newrelic(this: any, options: NewRelicOptions) {
+    const seneca: any = this
+
+    seneca
+      .message('plugin:newrelic,get:info', get_info)
+      .message('plugin:newrelic,api:metrics', metric_handler)
+      .message('plugin:newrelic,api:metrics', metrics_validation)
```

Comment:
don't define two actions with the same name - priors should be reserved mostly for user customisation